src/systems/world-update.js

```diff
@@ -0,0 +1,192 @@
+    //
+    // forceWorldUpdate - Forces the world matrix to be updated regardless of if the local matrix
+    // has been updated since the last update.
+    //
+    // skipParents - unless true, all parent matricies are updated before updating this object's
+    // local and world matrix.
+    //
+    // Returns true if the world matrix was updated.
+    //
+    THREE.Object3D.prototype.updateMatrices = function(forceLocalUpdate, forceWorldUpdate, skipParents) {
+      if (!this.hasHadFirstMatrixUpdate) {
+        if (
+          !this.position.equals(zeroPos) ||
+          !this.quaternion.equals(zeroQuat) ||
+          !this.scale.equals(oneScale) ||
+          !this.matrix.equals(identity)
+        ) {
+          // Only update the matrix the first time if its non-identity, this way
+          // this.matrixIsModified will remain false until the default
+          // identity matrix is updated.
+          this.updateMatrix();
+        }
+
+        this.hasHadFirstMatrixUpdate = true;
+        this.cachedMatrixWorld = this.matrixWorld;
+      } else if (this.matrixNeedsUpdate || this.matrixAutoUpdate || forceLocalUpdate) {
+        this.updateMatrix();
+        if (this.matrixNeedsUpdate) this.matrixNeedsUpdate = false;
+      }
+
+      if (!skipParents && this.parent) {
+        this.parent.updateMatrices(false, forceWorldUpdate, false);
+      }
+
+      if (this.matrixWorldNeedsUpdate || forceWorldUpdate) {
+        if (this.parent === null) {
+          this.matrixWorld.copy(this.matrix);
+        } else {
+          // If the matrix is unmodified, it is the identity matrix,
+          // and hence we can use the parent's world matrix directly.
+          //
+          // Note this assumes all callers will either not pass skipParents=true
+          // *or* will update the parent themselves beforehand as is done in
+          // updateMatrixWorld.
+          if (!this.matrixIsModified) {
+            this.matrixWorld = this.parent.matrixWorld;
+          } else {
+            // Once matrixIsModified === true, this.matrixWorld has been updated to be a local
+            // copy, not a reference to this.parent.matrixWorld (see updateMatrix/applyMatrix)
+            this.matrixWorld.multiplyMatrices(this.parent.matrixWorld, this.matrix);
+          }
+        }
+
+        this.matrixWorldNeedsUpdate = false;
+
+        return true;
+      }
+
+      return false;
+    };
+
+    // Computes this object's matrices and then the recursively computes the matrices of all the children.
+    //
+    // forceWorldUpdate - If true and the object is visible, will force the world matrix to be updated for
+    // this node and all of its children.
+    //
+    // includeInvisible - If true, does not ignore non-visible objects.
+    THREE.Object3D.prototype.updateMatrixWorld = function(forceWorldUpdate, includeInvisible) {
+      if (!this.visible && !includeInvisible) return;
+
+      // Do not recurse upwards, since this is recursing downwards
```

Comment:
confused by this comment. `this.updateMatricies` only recurses upwards no? What is the "this" in "this is only recursing downwards"

Reply:
`updateMatrixWorld` is recursing downwards, so if we passed `skipParents = false` to `updateMatrices` we'd:

- break the semantics of `updateMatrixWorld` inherited from three.js where there are no side effects on parents

- end up redundantly visiting each node an additional log(N) times (I think?)